templates/puppet/integration_spec.rb

```diff
@@ -0,0 +1,194 @@
+those same HTTP requests with those same responses, making your tests
+repeatable (desirable), fast (desirable), and in our case, also free!
+
+You can't write a test like that using the puppet command line tool, so
+instead we plumb into the guts of puppet.  Our main hook is a function called
+apply_compiled_manifest (and its sibling, apply_with_error_check).
+
+These functions are not really very similar to `puppet apply`.  Instead of
+taking a filename, they take the entire manifest as a string.  This causes
+subtle problems deep within Puppet - Puppet depends on the filename of the
+manifest in a handful of ways, not least of which is to find out what
+module it is part of (to automagically load dependencies and functions).
+Since that isn't possible, we tack on the Puppet override of our own
+environment and loader.  Since these functions don't call out to Facter
+the way that `puppet apply` does, we need to substitute in our own project
+name and credential path (which all our examples use).  Hence the
+get_example() function - it does more than just read the file.
+
+There's also the issue of the begin/rescue pattern you'll see here.  It's
+ugly!  It's also necessary.  compile_to_ral() calls out to our mocked up
+loader, but fails to successfully compile the manifest due to some
+initialization which is mandatory, but isn't possible to complete from
+outside puppet.  HOWEVER!  We're in luck!  In calling some cleanup code,
+in reporting on the failure, compile_to_ral actually *performs* the
+initialization we need.
+
+So!  We *can* successfully apply a manifest from
+under webmock, as long as we:
+- try twice, eating the first failure.
+- don't need to use any functions that are loaded from modules
+- ensure all our dependent modules have their 'lib' dir in $LOAD_PATH
+- do all our own variable substitution
+- rewrite any necessary functions in puppet and stick them in
+    the manifest.
+- import all the magic from puppetlabs_spec_helper/module_spec_helper,
+    which I have done without fully understanding it.  The comments in
+    the code suggest that the authors also do not fully understand it.
+
+Hopefully that explains the complexity here in a way that will continue
+to be useful as the tests develop further.
+
+Let me also explain the order in which these commands are run.  First,
+there is a pre-destroy - in the event that an older test failed to clean
+up after itself, the next run should take care of it.  Second, we create
+the resources in the manifest.  Third, we run the create again, mocking
+out any requests during 'flush' and making them crash the test - this is
+to check idempotency of the manifests.  Fourth, we delete the resources
+in the manifest, and fifth, we check the idempotency of that delete by
+confirming that no resources make web requests during 'flush'.
+
+VCR recordings are called 'cassettes', and they are stored in
+'spec/cassettes'.  If present, they are used - if absent, created.
```

Comment:
Are we planning on storing the cassettes on github and updating them periodically?

Reply:
Planning instead on storing them somewhere auth'd, because they include the raw text of the request we send with our service account private key.  But that's not part of this PR just yet - I want to get the tests in first, then integrate them.